internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,635 @@
+// { v1.x, v2.x }, {v1.z.x, v2.z.x }, ...
+func (p *transformer) unifyByValue(ref1 Reference, ref2 Reference) {
+	// A field may be addressable or non-addressable:
+	// (1) for the non-addressable case such that r1 -> r1.x and r2 -> r2.x, unify
+	// r1.x and r2.x;
+	// (2) for the addressable case such as r1 -> &r1.x -> *r1.x and r2 -> &r2.x
+	// -> *r2.x, unify *r1.x and *r2.x, and keep &r1.x and &r2.x intact.
+	state := p.state
+	rep1, rep2 := state.representative(ref1), state.representative(ref2)
+	fmap1, fmap2 := state.PartitionFieldMap(rep1), state.PartitionFieldMap(rep2)
+	if fmap1 == nil || fmap2 == nil {
+		log.Fatal("field maps are nil")
+	}
+	if &fmap1 == &fmap2 { // already unified
+		return
+	}
+	// The to-be-unified field pairs.
+	toUnified := make(map[Reference]Reference)
+	// "Copy" the fields from ref2 to ref1: copy only the field values rather than
+	// the field addresses.
+	for k, addr2 := range fmap2 {
+		var addr1 Reference
+		if fd, ok := fmap1[k]; ok {
+			addr1 = fd
+		} else {
+			addr1 = state.Insert(MakeSynthetic(1 /*FIELD*/, rep1))
+			fmap1[k] = addr1
+		}
+		toUnified[addr1] = addr2
+		// TODO: Recursively call "UnifyByValue" here.
+	}
+	// "Copy" the remaining fields in ref1 to ref2. Only copy unify-by-reference
+	// field values.
+	for k, addr1 := range fmap1 {
+		if _, ok := fmap2[k]; ok {
+			continue
+		}
+		addr2 := state.Insert(MakeSynthetic(1 /*FIELD*/, rep2))
+		fmap2[k] = addr2
+		toUnified[addr1] = addr2
+	}
+	// Perform further unifications.
+	addressable := isFieldAddressable(rep2.Type())
+	for addr1, addr2 := range toUnified {
+		p.unifyFields(addr1, addr2, addressable)
+	}
+}
+
+// Return whether a type can use the unify-by-reference semantics. If not then
+// it is assumed to be unifyByValue.
+func isUnifyByReference(tp types.Type) bool {
+	switch t := tp.(type) {
+	case *types.Struct,
+		*types.Array:
+		return false
+	case *types.Named:
+		return isUnifyByReference(t.Underlying())
+	}
+	// "MayShareObject" determines whether the value may share an object
+	// pointed by other values. It includes structs and arrays, whose elements may be shared.
+	return typeMayShareObject(tp)
+}
+
+// Return whether the fields of a type can be accessed through operator &
+// (https://golang.org/ref/spec#Address_operators).
+// The Go frontend and SSA handle an addressable field by introducing an
+// intermediate register for the address. For example, for struct A, "A.x = 1"
+// is compiled into "t1 = &A.x; *t1 = 1".
+func isFieldAddressable(tp types.Type) bool {
+	switch t := tp.(type) {
+	case *types.Struct,
+		*types.Slice,
+		*types.Array:
+		return true
+	case *types.Named:
+		return isFieldAddressable(t.Underlying())
+	}
+	return false
+}
+
+// Return the name of a field in a struct.
+// It the field is not found, then it returns the string format of the index.
+//
+// For example, for struct T {x int, y int), getFieldName(*T, 1) returns "y".
+func getFieldName(tp types.Type, index int) string {
```

Comment:
When "tp" is a proper struct type, then the field shall be located. Otherwise the field index is used instead. For example, 

type T struct {x int; y int}

getField(T, 1) will return Field{Name: "y", irField: y}. In case of something wrong, then it returns Field{Name: "1"}.